h/static/scripts/base/controller.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+var dom = require('../util/dom');
+
+/**
+ * Base class for controllers that upgrade elements with additional
+ * functionality.
+ *
+ * - Child elements with `js-$name` class prefixes are exposed on the controller
+ *   as `this.refs.$name` where $name is camelCased.
+ * - The element passed to the controller is exposed via the `element` property
+ * - The controllers attached to an element are accessible via the
+ *   `element.controllers` array
+ *
+ * The controller maintains internal state in `this.state`, which can only be
+ * updated by calling (`this.setState(changes)`). Whenever the internal state of
+ * the controller changes, `this.update()` is called to sync the DOM with this
+ * state.
+ *
+ * @param {Element} element - The DOM Element to upgrade
+ */
+function Controller(element) {
+  if (!element.controllers) {
+    element.controllers = [this];
+  } else {
+    element.controllers.push(this);
+  }
+
+  this.state = {};
+  this.element = element;
+  this.refs = dom.jsElements(element);
+}
+
+/**
+ * Set the state of the controller.
+ *
+ * This will merge `changes` into the current state and call the `update()`
+ * method provided by the subclass to update the DOM to match the current state.
+ */
+Controller.prototype.setState = function (changes) {
```

Comment:
It might be better if this were called `updateState()` or maybe `mergeState()`? To me `set` implies that `state` is completely replaced with the new state that's passed in.